runtime/cudaq/platform/default/rest_server/helpers/RestRemoteClient.cpp

```diff
@@ -0,0 +1,223 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "common/RuntimeMLIR.h"
+#include "cudaq/Frontend/nvqpp/AttributeNames.h"
+#include "cudaq/Optimizer/CodeGen/OpenQASMEmitter.h"
+#include "cudaq/Optimizer/CodeGen/Passes.h"
+#include "cudaq/Optimizer/Dialect/CC/CCDialect.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeDialect.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+
+#include "llvm/Bitcode/BitcodeReader.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/ExecutionEngine/ExecutionEngine.h"
+#include "mlir/ExecutionEngine/OptUtils.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Pass/PassRegistry.h"
+
+#include "JsonConvert.h"
+#include "common/Logger.h"
+#include "common/RemoteKernelExecutor.h"
+#include "common/RestClient.h"
+#include "cudaq.h"
+
+namespace {
+using namespace mlir;
+class RemoteRestRuntimeClient : public cudaq::RemoteRuntimeClient {
+  std::string m_url;
+  static inline const std::vector<std::string> clientPasses = {
+      "func.func(unwind-lowering)",
+      "func.func(indirect-to-direct-calls)",
+      "inline",
+      "canonicalize",
+      "apply-op-specialization",
+      "func.func(memtoreg{quantum=0})",
+      "canonicalize",
+      "expand-measurements",
+      "cc-loop-normalize",
+      "cc-loop-unroll",
+      "canonicalize",
+      "func.func(add-dealloc)",
+      "func.func(quake-add-metadata)",
+      "canonicalize",
+      "func.func(lower-to-cfg)",
+      "func.func(combine-quantum-alloc)",
+      "canonicalize",
+      "cse",
+      "quake-to-qir"};
```

Comment:
This is a nice way to split some processing between the client and the server, but in the future, we may want to look at how to combine this with `addPipelineToQIR` so that we use common code for all these places that do passes lest the pass pipeline differences become unmaintainable. (No change needed for this PR unless you already have a good idea on how to do that.)

Reply:
Agreed. I don't currently have a good idea :) Maybe we should add a ticket to track this?